add read-command subcommand

blu-nitro opened the pull request:
- Adds read-command subcommand that reads ci commands from a json file generated by a GitHub webhook for issue_comment and triggers gitlab ci pipelines for them
- Moves original main command to write subcommand

Todo:
- [x] Add unit tests
- [x] Test with real repositories
- [x] Add possible bot names as argument
- [x] Add bot name as GitLab CI variable sent with pipeline trigger

blu-nitro committed:
add trigger-webhook-commands subcommand

this commnand triggers CI pipelines with commands from webhook

robin-nitrokey reviewed (commented):
Thanks for the PR!  I would prefer to have the webhook parsing in a separate module together with some unit test that parse valid and invalid webhook content.

Generally, I think `@<bot> commands...` is the common format for this kind of commands on GitHub.  AFAIS this would use `!commands...` instead.  Is there a special reason for that?

blu-nitro commented:
I refactored a few things, but tests are yet to come.

> Generally, I think `@<bot> commands...` is the common format for this kind of commands on GitHub. AFAIS this would use `!commands...` instead. Is there a special reason for that?

Honestly the command trigger was just one i remembered seeing on GitHub lately and it was more of a placeholder for now. I changed it to `@bot commands...` now and you could even have multiple commands in multiple lines like this. But i still wanted to retain the possibility to add a comment with lines that don't start with `@bot`.

robin-nitrokey commented on src/args.rs:
```diff
@@ -1,37 +1,64 @@
+    ///
+    /// The GITHUB_COMMENT_TOKEN environment variable must be set to a token with the write permission
+    /// for issues or pull requests on the target repository.
+    Write(WriteArgs),
```
Maybe `WriteComment` to be more specific?

blu-nitro replied:
done

robin-nitrokey commented on src/args.rs:
```diff
@@ -1,37 +1,64 @@
+    ///
+    /// The GITHUB_COMMENT_TOKEN environment variable must be set to a token with the read permission
+    /// for issues or pull requests on the target repository.
+    ReadCommand(ReadCommandArgs),
```
The description is outdated.  As the command not only reads the commands but also submits them to the CI, I think something like `ExecCommands`/`TriggerWebhookCommands`/... would be a better name.

blu-nitro replied:
done

robin-nitrokey commented on src/webhook.rs:
```diff
@@ -0,0 +1,345 @@
+mod tests {
+    use super::*;
+
+    const WEBHOOK_DATA: &str = r#"
```
Could also be moved to a separate file and [`include_str!`](https://doc.rust-lang.org/std/macro.include_str.html)-ed.

blu-nitro replied:
Like this? I'm not sure that in src is the right place for this test data, or is it?

robin-nitrokey replied:
Yes, like this.  I would just put it in a `test` or `data` subdirectory.

blu-nitro replied:
done

robin-nitrokey commented on src/webhook.rs:
```diff
@@ -0,0 +1,345 @@
+
+    #[test]
+    fn test_parse_webhook() {
+        let webhook = Webhook::parse_from_str(WEBHOOK_DATA).unwrap();
```
Test for `webhook.comment` is missing.

blu-nitro replied:
done

robin-nitrokey reviewed (commented):
This feature is very specific to our CI setup so I would rename the tool from `github-comment` to something like `nitrokey-ci`.  What do you think?

blu-nitro commented:
> This feature is very specific to our CI setup so I would rename the tool from `github-comment` to something like `nitrokey-ci`. What do you think?

That is true, it is going beyond just acting on comments now. Although it can be used with any GitLab instance and any repository in just the same way, as long as the repository path is the same on GitHub and GitLab. But a name that represents all of that would get very verbose. So i would stick with `nitrokey-ci` and possibly at a later point explain how to use this tool in other scenarios in a readme.

blu-nitro committed:
move test data

robin-nitrokey commented on src/read_command.rs:
```diff
@@ -0,0 +1,88 @@
+    let mut bot_commands: Vec<BotCommand> = Vec::new();
+    for bot in bots {
+        let atbot = &format!("@{bot} ");
+        text.lines()
+            .filter(|line| line.starts_with(atbot))
+            .flat_map(|s| -> Vec<String> {
+                s.to_string()
+                    .replace(atbot, "")
+                    .split(' ')
+                    .map(|s| s.to_string())
+                    .collect()
+            })
```
You can keep this as it is, but I just wanted to mention that there is a cool function [`str::strip_prefix`](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix) that could be useful here.  Also, it is most often not necessary to `collect` inside chained iterators:

```suggestion
        text.lines()
            .flat_map(|line| line.strip_prefix(atbot))
            .flat_map(|s| s.split(' '))
```

The first `flat_map` filters for lines starting with `atbot` and removes it at the same time.  The second `flat_map` replaces the `collect` and produces a single iterator over the commands.

blu-nitro replied:
This is a lot nicer thank you!
Also yes, as this was supposed to be as general purpose as possible, i think you are right in suggesting to add arguments to commands. This is added now.

robin-nitrokey reviewed (commented):
Two final remarks before we can merge:
- With `read_command` being renamed to `trigger_webhook_commands` the `read_command` module and `read` function should be renamed too.
- Wouldn’t it make sense to treat a single line as a single command instead of splitting at spaces?  For example, I could imagine commands with additional arguments like `@bot run-ci slow-job-42`.

blu-nitro committed:
rename

blu-nitro committed:
change multiple commands per line to one with arguments

robin-nitrokey commented on src/webhook_command_utils.rs:
```diff
@@ -0,0 +1,77 @@
+        let atbot = &format!("@{bot} ");
+        text.lines()
+            .flat_map(|line| line.strip_prefix(atbot))
+            .flat_map(|line| line.split_once(' '))
```
I think this misses the case of a command without arguments (should also be added to the tests).

blu-nitro replied:
Yes i caught that in CI as well

blu-nitro committed:
fix commands with no argument are still commands

blu-nitro committed:
add test case for command without args

robin-nitrokey reviewed (approved):
Thanks for the fixes.  LGTM!
